README.md
node-decoder-ring
=================

[![Build Status](https://travis-ci.org/asbaker/node-decoder-ring.png)](https://travis-ci.org/asbaker/node-decoder-ring)

Decoder Ring allows you to use a Javascript object as a specification to decode [Node.js Buffers](http://nodejs.org/api/buffer.html) into a Javascript object.

## Installation

	npm install decoder-ring

##  Usage


### Javascript Object Specification

The Javascript object specification is used to specify endianness and a description of the fields present in the buffer.
```javascript
{
    bigEndian: true,
    fields: [
        {name: "field1", start: 0,  type: 'int8'  },
        {name: "field2", start: 1,  type: 'uint8' },
        {name: "field3", start: 2,  type: 'int16' },
        {name: "field4", start: 4,  type: 'uint16'},
        {name: "field5", start: 6,  type: 'float' },
        {name: "field6", start: 10, type: 'double'},
        {name: "field7", start: 18, type: 'ascii', length: 10 },
        {name: "field8", start: 28, type: 'utf8',  length: 9  },
        {name: "field9", start: 37, type: 'bit', position: 7},
        {name: "field10", start: 37, type: 'bit', position: 6},
        {name: "field11", start: 37, type: 'bit', position: 0},
        {name: "field12", start: 38, type: 'uint32'},
        {name: "field13", start: 42, type: 'int32'}
        {name: "field14", start: 46, type: 'int8', default: 42 }
        {name: "field15", start: 50, type: 'buffer', length: 4 }
    ]
}


```

All fields must have a name, a starting byte, and a type. The name is used for assigning the property in the resulting javascript object. Additionally, fields can have a default value which is used when encoding an object's properties that have null or undefined values.

#### Types
* **buffer** - Raw buffer
* **int8** - Signed 8-bit integer
* **uint8** - Unsigned 8-bit integer
* **int16** - Signed 16-bit integer
* **uint16** - Unsigned 16-bit integer
* **int32** - Signed 32-bit integer
* **uint32** - Unsigned 32-bit integer
* **float** - 4-bit floating point number
* **double** - 8-bit double precision floating point number
* **ascii** - 8-bit per character ASCII encoded text

	This field type must also have a length property which is a count of the number of characters.
* **utf8** - 8-bit per character UTF8 encoded text

	This field type must also have a length property which is a count of the number of characters.
* **bit** - true/false values

	Bit fields are pieces of a 1-byte unsigned integer. Given a big endian, unsigned integer of 129, it will appear as the following when broken down into bits:


  | 128(2^7) | 64(2^6) | 32(2^5) | 16(2^4) | 8(2^3) | 4(2^2) | 2(2^1) | 1(2^0) |
  | :--:	 | :--:	| :--:	| :--:	| :--:   | :--:   | :--:   | :--:   |
  | 1		| 0	   | 0	   | 0	   | 0	  | 0	  | 0	  | 1	  |

	Bit fields must have a position property which is used to check if a specific bit is on or off in the 1-byte unsigned integer.
	The position of the bit of interest, is defined as which power of two the bit falls in the integer. For the bit in the 128th's place, the position would be 7, for the bit in the 1's place the position would be 0.

### API

decode: (buffer, spec, { noAssert = false })
  * set noAssert to true to ignore validation of offsets

encode: (obj, spec, { noAssert = false, padding = null })
  * set padding to a character to right pad ascii and utf8 strings instead of using null termination
  * set noAssert to true to ignore validation of written values and offsets

### Example

```javascript
var DecoderRing = require("decoder-ring");
var decoderRing = new DecoderRing();

var bufferBE = Buffer.alloc(51);
bufferBE.writeInt8(-127, 0);
bufferBE.writeUInt8(254, 1);
bufferBE.writeInt16BE(5327, 2);
bufferBE.writeUInt16BE(5328, 4);
bufferBE.writeFloatBE(-15.33, 6);
bufferBE.writeDoubleBE(-1534.98, 10);
bufferBE.write("ascii", 18, 10,'ascii');
bufferBE.write("utf8 text", 28, 9, 'utf8');
bufferBE.writeUInt8(129, 37);
bufferBE.writeUInt32BE(79001, 38);
bufferBE.writeInt32BE(-79001, 42);
bufferBE.writeInt8(1, 46);

var testBuffer = new Buffer("test");
testBuffer.copy(bufferBE, 47, 0, 4);

var spec = {
  bigEndian: true,
  fields: [
    { name: "field1", start: 0,   type: 'int8'  },
    { name: "field2", start: 1,   type: 'uint8' },
    { name: "field3", start: 2,   type: 'int16' },
    { name: "field4", start: 4,   type: 'uint16'},
    { name: "field5", start: 6,   type: 'float' },
    { name: "field6", start: 10,  type: 'double'},
    { name: "field7", start: 18,  type: 'ascii', length: 10 },
    { name: "field8", start: 28,  type: 'utf8', length: 9  },
    { name: "field9", start: 37,  type: 'bit', position: 7 },
    { name: "field10", start: 37, type: 'bit', position: 6 },
    { name: "field11", start: 37, type: 'bit', position: 0 },
    { name: "field12", start: 38, type: 'uint32' },
    { name: "field13", start: 42, type: 'int32' },
    { name: "field14", start: 46, type: 'int8', default: 42 },
    { name: "field15", start: 47, type: 'buffer', length: 4 }
  ]
};

// Decode the buffer into a javascript object
var result = decoderRing.decode(bufferBE, spec);
console.log(result);

// Assign field14 to undefined to test default value on encoding
result.field14 = undefined;

// Encode the object to a buffer
var buffer = decoderRing.encode(result, spec);
console.log(buffer);

// Decode buffer to object and check field14 for default value
var resultWithDefaultValue = decoderRing.decode(buffer, spec);
console.log("Field14 default value: " + resultWithDefaultValue.field14);
```

Result of the above example

```javascript
{ 
  field1: -127,
  field2: 254,
  field3: 5327,
  field4: 5328,
  field5: -15.329999923706055,
  field6: -1534.98,
  field7: 'ascii\u0000\u0000\u0000\u0000\u0000',
  field8: 'utf8 text',
  field9: true,
  field10: false,
  field11: true,
  field12: 79001,
  field13: -79001,
  field14: 1,
  field15: <Buffer 74 65 73 74> 
}

<Buffer 81 fe 14 cf 14 d0 c1 75 47 ae c0 97 fb eb 85 1e b8 52 61 73 63 69 69 20 20 20 20 20 75 74 66 38 20 74 65 78 74 81 00 01 34 99 ff fe cb 67 2a>

Field14 42
```

## Development

### Running tests
	npm test

### Testing the package locally
	npm pack
	npm install decoder-ring-0.4.0.tgz
	node example.js




CHANGELOG.md
# CHANGELOG

## 1.0.0 (2017-04-20)
- Allow cached length to be included on specs
- Add noAssert and padding options
- Remove node 0.6.0 compatibility 
- Add yarn lockfile

## 0.5.0 (2015-05-28)
- Added raw buffer type

## 0.4.0 (2013-03-26)
- Added default value field

## 0.3.2 (2013-03-24)
- Performance improvements for field encoding, write to the same buffer when encoding fields

## 0.3.1 (2013-03-24)
- Add handling for undefined and null values in FieldEncoder

## 0.3.0 (2013-03-23)
- Add functions to encode javascript object to a buffer according to the specification
- Upgrade coffeescript devDependency to 1.6.2

## 0.2.0 (2013-02-26)
- Add int32 and uint32 field types, closes github issue #3
- Internal project restructuring
- Upgrade coffeescript devDependency to 1.5.0

## 0.1.0 (2013-02-24)
- Initial Release


lib/DecoderRing.js
// Generated by CoffeeScript 1.9.1
(function() {
  var DecoderRing, FieldDecoder, FieldEncoder;

  FieldDecoder = require("./FieldDecoder");

  FieldEncoder = require("./FieldEncoder");

  DecoderRing = (function() {
    function DecoderRing(fieldDecoder, fieldEncoder) {
      this.fieldDecoder = fieldDecoder != null ? fieldDecoder : new FieldDecoder;
      this.fieldEncoder = fieldEncoder != null ? fieldEncoder : new FieldEncoder;
    }

    DecoderRing.prototype.decode = function(buffer, spec) {
      var decodeFun, field, i, len, obj, ref;
      obj = {};
      if (spec.bigEndian) {
        decodeFun = this.fieldDecoder.decodeFieldBE;
      } else {
        decodeFun = this.fieldDecoder.decodeFieldLE;
      }
      ref = spec.fields;
      for (i = 0, len = ref.length; i < len; i++) {
        field = ref[i];
        obj[field.name] = decodeFun(buffer, field);
      }
      return obj;
    };

    DecoderRing.prototype.encode = function(obj, spec) {
      var bitFieldAccumulator, buffer, currentVal, encodeFun, fieldSpec, i, j, len, len1, r, ref, ref1, ref2, size, val;
      size = (ref = spec.length) != null ? ref : this.fieldEncoder.findSpecBufferSize(spec);
      buffer = new Buffer(size);
      buffer.fill(0);
      if (spec.bigEndian) {
        encodeFun = this.fieldEncoder.encodeFieldBE;
      } else {
        encodeFun = this.fieldEncoder.encodeFieldLE;
      }
      bitFieldAccumulator = {};
      ref1 = spec.fields;
      for (i = 0, len = ref1.length; i < len; i++) {
        fieldSpec = ref1[i];
        if (fieldSpec.type !== 'bit') {
          buffer = encodeFun(buffer, obj, fieldSpec);
        } else {
          val = obj[fieldSpec.name] ? Math.pow(2, fieldSpec.position) : 0;
          currentVal = bitFieldAccumulator["" + fieldSpec.start] || 0;
          bitFieldAccumulator["" + fieldSpec.start] = currentVal + val;
        }
      }
      ref2 = Object.keys(bitFieldAccumulator);
      for (j = 0, len1 = ref2.length; j < len1; j++) {
        r = ref2[j];
        buffer = encodeFun(buffer, bitFieldAccumulator, {
          name: r,
          start: parseInt(r),
          type: 'uint8'
        });
      }
      return buffer;
    };

    return DecoderRing;

  })();

  module.exports = DecoderRing;

}).call(this);


lib/FieldDecoder.js
// Generated by CoffeeScript 1.9.1
(function() {
  var FieldDecoder;

  FieldDecoder = (function() {
    function FieldDecoder() {}

    FieldDecoder.prototype.decodeFieldBE = function(buffer, fieldSpec) {
      var i;
      switch (fieldSpec.type) {
        case 'int8':
          return buffer.readInt8(fieldSpec.start);
        case 'uint8':
          return buffer.readUInt8(fieldSpec.start);
        case 'int16':
          return buffer.readInt16BE(fieldSpec.start);
        case 'uint16':
          return buffer.readUInt16BE(fieldSpec.start);
        case 'int32':
          return buffer.readInt32BE(fieldSpec.start);
        case 'uint32':
          return buffer.readUInt32BE(fieldSpec.start);
        case 'float':
          return buffer.readFloatBE(fieldSpec.start);
        case 'double':
          return buffer.readDoubleBE(fieldSpec.start);
        case 'ascii':
          return buffer.toString('ascii', fieldSpec.start, fieldSpec.start + fieldSpec.length);
        case 'utf8':
          return buffer.toString('utf8', fieldSpec.start, fieldSpec.start + fieldSpec.length);
        case 'buffer':
          return buffer.slice(fieldSpec.start, fieldSpec.start + fieldSpec.length);
        case 'bit':
          i = buffer.readUInt8(fieldSpec.start);
          return (i & Math.pow(2, fieldSpec.position)) > 0;
      }
    };

    FieldDecoder.prototype.decodeFieldLE = function(buffer, fieldSpec) {
      var i;
      switch (fieldSpec.type) {
        case 'int8':
          return buffer.readInt8(fieldSpec.start);
        case 'uint8':
          return buffer.readUInt8(fieldSpec.start);
        case 'int16':
          return buffer.readInt16LE(fieldSpec.start);
        case 'uint16':
          return buffer.readUInt16LE(fieldSpec.start);
        case 'int32':
          return buffer.readInt32LE(fieldSpec.start);
        case 'uint32':
          return buffer.readUInt32LE(fieldSpec.start);
        case 'float':
          return buffer.readFloatLE(fieldSpec.start);
        case 'double':
          return buffer.readDoubleLE(fieldSpec.start);
        case 'ascii':
          return buffer.toString('ascii', fieldSpec.start, fieldSpec.start + fieldSpec.length);
        case 'utf8':
          return buffer.toString('utf8', fieldSpec.start, fieldSpec.start + fieldSpec.length);
        case 'buffer':
          return buffer.slice(fieldSpec.start, fieldSpec.start + fieldSpec.length);
        case 'bit':
          i = buffer.readUInt8(fieldSpec.start);
          return (i & Math.pow(2, fieldSpec.position)) > 0;
      }
    };

    return FieldDecoder;

  })();

  module.exports = FieldDecoder;

}).call(this);


lib/FieldEncoder.js
// Generated by CoffeeScript 1.9.1
(function() {
  var FieldEncoder;

  FieldEncoder = (function() {
    function FieldEncoder() {}

    FieldEncoder.prototype.findSpecBufferSize = function(spec) {
      var f, i, len, ref, sizes;
      sizes = [];
      ref = spec.fields;
      for (i = 0, len = ref.length; i < len; i++) {
        f = ref[i];
        sizes.push(this.findFieldLength(f));
      }
      return sizes.reduce(function(a, b) {
        return Math.max(a, b);
      });
    };

    FieldEncoder.prototype.findFieldLength = function(field) {
      var length;
      length = (function() {
        switch (field.type) {
          case 'int8':
          case 'uint8':
          case 'bit':
            return 1;
          case 'int16':
          case 'uint16':
            return 2;
          case 'uint32':
          case 'int32':
          case 'float':
            return 4;
          case 'double':
            return 8;
          case 'ascii':
          case 'utf8':
          case 'buffer':
            return field.length;
        }
      })();
      return field.start + length;
    };

    FieldEncoder.prototype.encodeFieldBE = function(buffer, obj, fieldSpec) {
      var val;
      val = obj[fieldSpec.name];
      if ((val == null) && (fieldSpec["default"] != null)) {
        val = fieldSpec["default"];
      }
      if (val != null) {
        switch (fieldSpec.type) {
          case 'int8':
            buffer.writeInt8(val, fieldSpec.start);
            break;
          case 'uint8':
            buffer.writeUInt8(val, fieldSpec.start);
            break;
          case 'int16':
            buffer.writeInt16BE(val, fieldSpec.start);
            break;
          case 'uint16':
            buffer.writeUInt16BE(val, fieldSpec.start);
            break;
          case 'float':
            buffer.writeFloatBE(val, fieldSpec.start);
            break;
          case 'double':
            buffer.writeDoubleBE(val, fieldSpec.start);
            break;
          case 'ascii':
            buffer.write(val, fieldSpec.start, fieldSpec.length, 'ascii');
            break;
          case 'utf8':
            buffer.write(val, fieldSpec.start, fieldSpec.length, 'utf8');
            break;
          case 'uint32':
            buffer.writeUInt32BE(val, fieldSpec.start);
            break;
          case 'int32':
            buffer.writeInt32BE(val, fieldSpec.start);
            break;
          case 'buffer':
            val.copy(buffer, fieldSpec.start, 0, fieldSpec.length);
            break;
          case 'bit':
            if (val === true) {
              buffer.writeUInt8(Math.pow(2, fieldSpec.position), fieldSpec.start);
            } else {
              buffer.writeUInt8(0, fieldSpec.start);
            }
        }
      }
      return buffer;
    };

    FieldEncoder.prototype.encodeFieldLE = function(buffer, obj, fieldSpec) {
      var val;
      val = obj[fieldSpec.name];
      if ((val == null) && (fieldSpec["default"] != null)) {
        val = fieldSpec["default"];
      }
      if (val != null) {
        switch (fieldSpec.type) {
          case 'int8':
            buffer.writeInt8(val, fieldSpec.start);
            break;
          case 'uint8':
            buffer.writeUInt8(val, fieldSpec.start);
            break;
          case 'int16':
            buffer.writeInt16LE(val, fieldSpec.start);
            break;
          case 'uint16':
            buffer.writeUInt16LE(val, fieldSpec.start);
            break;
          case 'float':
            buffer.writeFloatLE(val, fieldSpec.start);
            break;
          case 'double':
            buffer.writeDoubleLE(val, fieldSpec.start);
            break;
          case 'ascii':
            buffer.write(val, fieldSpec.start, fieldSpec.length, 'ascii');
            break;
          case 'utf8':
            buffer.write(val, fieldSpec.start, fieldSpec.length, 'utf8');
            break;
          case 'uint32':
            buffer.writeUInt32LE(val, fieldSpec.start);
            break;
          case 'int32':
            buffer.writeInt32LE(val, fieldSpec.start);
            break;
          case 'buffer':
            val.copy(buffer, fieldSpec.start, 0, fieldSpec.length);
            break;
          case 'bit':
            if (val === true) {
              buffer.writeUInt8(Math.pow(2, fieldSpec.position), fieldSpec.start);
            } else {
              buffer.writeUInt8(0, fieldSpec.start);
            }
        }
      }
      return buffer;
    };

    return FieldEncoder;

  })();

  module.exports = FieldEncoder;

}).call(this);


example.js
var DecoderRing = require("decoder-ring");
var decoderRing = new DecoderRing();

var bufferBE = Buffer.alloc(51);
bufferBE.writeInt8(-127, 0);
bufferBE.writeUInt8(254, 1);
bufferBE.writeInt16BE(5327, 2);
bufferBE.writeUInt16BE(5328, 4);
bufferBE.writeFloatBE(-15.33, 6);
bufferBE.writeDoubleBE(-1534.98, 10);
bufferBE.write("ascii", 18, 10,'ascii');
bufferBE.write("utf8 text", 28, 9, 'utf8');
bufferBE.writeUInt8(129, 37);
bufferBE.writeUInt32BE(79001, 38);
bufferBE.writeInt32BE(-79001, 42);
bufferBE.writeInt8(1, 46);

var testBuffer = new Buffer("test");
testBuffer.copy(bufferBE, 47, 0, 4);

var spec = {
  bigEndian: true,
  fields: [
    { name: "field1", start: 0,   type: 'int8'  },
    { name: "field2", start: 1,   type: 'uint8' },
    { name: "field3", start: 2,   type: 'int16' },
    { name: "field4", start: 4,   type: 'uint16'},
    { name: "field5", start: 6,   type: 'float' },
    { name: "field6", start: 10,  type: 'double'},
    { name: "field7", start: 18,  type: 'ascii', length: 10 },
    { name: "field8", start: 28,  type: 'utf8', length: 9  },
    { name: "field9", start: 37,  type: 'bit', position: 7 },
    { name: "field10", start: 37, type: 'bit', position: 6 },
    { name: "field11", start: 37, type: 'bit', position: 0 },
    { name: "field12", start: 38, type: 'uint32' },
    { name: "field13", start: 42, type: 'int32' },
    { name: "field14", start: 46, type: 'int8', default: 42 },
    { name: "field15", start: 47, type: 'buffer', length: 4 }
  ]
};

// Decode the buffer into a javascript object
var result = decoderRing.decode(bufferBE, spec);
console.log(result);

// Assign field14 to undefined to test default value on encoding
result.field14 = undefined;

// Encode the object to a buffer
var buffer = decoderRing.encode(result, spec);
console.log(buffer);

// Decode buffer to object and check field14 for default value
var resultWithDefaultValue = decoderRing.decode(buffer, spec);
console.log("Field14 default value: " + resultWithDefaultValue.field14);

